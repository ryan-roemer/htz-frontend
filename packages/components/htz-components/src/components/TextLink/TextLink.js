import React from 'react';
import PropTypes from 'prop-types';
import { FelaComponent, } from 'react-fela';
import { parseStyleProps, } from '@haaretz/htz-css-tools';
import HtzLink from '../HtzLink/HtzLink';
import { stylesPropType, } from '../../propTypes/stylesPropType';

const inlineLinkStyle = ({
  theme: { color, articleStyle, },
  theme,
  miscStyles,
}) => {
  const focusActiveStyles = {
    color: color('link', 'base'),
    borderBottomColor: color('link', 'base'),
  };
  return {
    borderBottomColor: color('link', 'base'),
    borderBottomWidth: articleStyle.paragraphLink.borderBottomWidth,
    borderBottomStyle: articleStyle.paragraphLink.borderBottomStyle,
    transitionProperty: 'all',

    ':hover': {
      ...focusActiveStyles,
      borderBottomColor: 'transparent',
    },
    ':focus': focusActiveStyles,
    ':active': focusActiveStyles,
    ':visited': { color: color('bodyText', 'base'), },
    extend: [
      theme.getTransition(0, 'swiftOut'),
      ...(miscStyles ? parseStyleProps(miscStyles, theme.mq, theme.type) : []),
    ],
  };
};

TextLink.propTypes = {
  /** The HTML tag to render the `<Button />` as */
  tagName: PropTypes.string,
  /**
   * A special property holding miscellaneous CSS values that
   * trump all default values. Processed by
   * [`parseStyleProps`](https://Haaretz.github.io/htz-frontend/htz-css-tools#parsestyleprops)
   */
  miscStyles: stylesPropType,
};
TextLink.defaultProps = {
  tagName: null,
  miscStyles: null,
};

function TextLink({ tagName, miscStyles, ...props }) {
  const Tag = tagName || HtzLink;
  return (
    <FelaComponent
      miscStyles={miscStyles}
      rule={inlineLinkStyle}
      render={({ className, }) => <Tag className={className} {...props} />}
    />
  );
}

export default TextLink;
