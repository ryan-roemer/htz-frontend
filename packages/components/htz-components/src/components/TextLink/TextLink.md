### **Text Link component**

`<TextLink />` is a ui component that styles our links.

By default text link will render `<HtzLink/>` component.

Passing a tagName prop will override using `<HtzLink/>`

All other props are passed to the rendered Element

Render as a default HtzLink

```jsx
<TextLink href="https://www.haaretz.co.il" content="Haaretz" />
```

Render as a button

```jsx
<TextLink
  tagName="button"
  href="https://www.haaretz.co.il"
  onClick={() => console.log('I can get an onClick')}>
  Haaretz
</TextLink>
```
