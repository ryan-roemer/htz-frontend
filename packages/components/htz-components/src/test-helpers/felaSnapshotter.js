import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
