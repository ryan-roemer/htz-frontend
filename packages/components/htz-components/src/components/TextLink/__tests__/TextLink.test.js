import React from 'react';
import { ApolloProvider, } from 'react-apollo';
import client from '../../../../styleguide/ApolloMockClient';
import felaSnapshotter from '../../../test-helpers/felaSnapshotter';
import TextLink from '../TextLink';

describe('<TextLink>', () => {
  describe('DOM element', () => {
    it('renders correctly with minimal required props', () => {
      const { component, styles, } = felaSnapshotter(
        <ApolloProvider client={client}>
          <TextLink href="test.com">I am a htzLink</TextLink>
        </ApolloProvider>
      );
      expect(component).toMatchSnapshot();
      expect(styles).toMatchSnapshot();
    });
    it('renders correctly with miscStyles ,', () => {
      const { component, styles, } = felaSnapshotter(
        <ApolloProvider client={client}>
          <TextLink href="test.com" miscStyles={{ color: 'blue', }}>
            I am a htzLink
          </TextLink>
        </ApolloProvider>
      );
      expect(component).toMatchSnapshot();
      expect(styles).toMatchSnapshot();
    });
    it('renders correctly with cusom tagName ,', () => {
      const { component, styles, } = felaSnapshotter(
        <TextLink tagName="button">I am a button</TextLink>
      );
      expect(component).toMatchSnapshot();
      expect(styles).toMatchSnapshot();
    });
  });
});
