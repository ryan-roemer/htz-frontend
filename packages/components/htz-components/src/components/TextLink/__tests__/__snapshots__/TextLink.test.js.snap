// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextLink> DOM element renders correctly with cusom tagName , 1`] = `
<div>
  <button
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n"
  >
    I am a button
  </button>
</div>
`;

exports[`<TextLink> DOM element renders correctly with cusom tagName , 2`] = `".htz-a{color:#0B7EB5}.htz-b:visited{color:#0B7EB5}.htz-c:hover{color:#0B7EB5}.htz-d:hover{border-bottom-color:#0B7EB5}.htz-e:hover{border-bottom-width:1px}.htz-f:hover{border-bottom-style:solid}.htz-g:focus{color:#0B7EB5}.htz-h:focus{border-bottom-color:#0B7EB5}.htz-i:focus{border-bottom-width:1px}.htz-j:focus{border-bottom-style:solid}.htz-k:active{color:#0B7EB5}.htz-l:active{border-bottom-color:#0B7EB5}.htz-m:active{border-bottom-width:2px}.htz-n:active{border-bottom-style:solid}"`;

exports[`<TextLink> DOM element renders correctly with minimal required props 1`] = `
<div>
  <a
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n"
    href="test.com"
    onClick={null}
  >
    I am a htzLink
  </a>
</div>
`;

exports[`<TextLink> DOM element renders correctly with minimal required props 2`] = `".htz-a{color:#0B7EB5}.htz-b:visited{color:#0B7EB5}.htz-c:hover{color:#0B7EB5}.htz-d:hover{border-bottom-color:#0B7EB5}.htz-e:hover{border-bottom-width:1px}.htz-f:hover{border-bottom-style:solid}.htz-g:focus{color:#0B7EB5}.htz-h:focus{border-bottom-color:#0B7EB5}.htz-i:focus{border-bottom-width:1px}.htz-j:focus{border-bottom-style:solid}.htz-k:active{color:#0B7EB5}.htz-l:active{border-bottom-color:#0B7EB5}.htz-m:active{border-bottom-width:2px}.htz-n:active{border-bottom-style:solid}"`;

exports[`<TextLink> DOM element renders correctly with miscStyles , 1`] = `
<div>
  <a
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n"
    href="test.com"
    onClick={null}
  >
    I am a htzLink
  </a>
</div>
`;

exports[`<TextLink> DOM element renders correctly with miscStyles , 2`] = `".htz-a{color:blue}.htz-b:visited{color:#0B7EB5}.htz-c:hover{color:#0B7EB5}.htz-d:hover{border-bottom-color:#0B7EB5}.htz-e:hover{border-bottom-width:1px}.htz-f:hover{border-bottom-style:solid}.htz-g:focus{color:#0B7EB5}.htz-h:focus{border-bottom-color:#0B7EB5}.htz-i:focus{border-bottom-width:1px}.htz-j:focus{border-bottom-style:solid}.htz-k:active{color:#0B7EB5}.htz-l:active{border-bottom-color:#0B7EB5}.htz-m:active{border-bottom-width:2px}.htz-n:active{border-bottom-style:solid}"`;
